lib/authorizer/hawk.js

```diff
@@ -24,6 +28,84 @@ function randomString (length) {
     return result.join(EMPTY);
 }
 
+/**
+ * Encodes characters not encoded by querystring.stringify() according to RFC3986.
+ * REFER: https://github.com/postmanlabs/postman-request/blob/master/lib/querystring.js#L42
+ *
+ * @param {String} str Partially encoded string by querystring.stringify()
+ * @returns {String} Fully encoded string
+ */
+function rfc3986 (str) {
+    return str.replace(/[!'()*]/g, function (c) {
+        return '%' + c.charCodeAt(0).toString(16).toUpperCase();
+    });
+}
+
+/**
+ * Calculates body hash with given algorithm and digestEncoding.
+ * REFER: https://github.com/postmanlabs/postman-request/blob/master/lib/hawk.js#L12
+ *
+ * @param {RequestBody} body
+ * @param {String} algorithm
+ * @param {String} digestEncoding
+ * @param {String} contentType
+ * @param {Function} callback
+ */
+function computeBodyHash (body, algorithm, digestEncoding, contentType, callback) {
+    if (!(body && algorithm && digestEncoding) || body.isEmpty()) { return callback(); }
+
+    var hash = crypto.createHash(algorithm),
+        originalReadStream,
+        urlencodedBody;
+
+    hash.update('hawk.1.payload\n');
+    hash.update((contentType ? contentType.split(';')[0].trim().toLowerCase() : '') + '\n');
+
+    if (body.mode === RequestBody.MODES.raw) {
+        hash.update(body.toString());
+        hash.update('\n');
+
+        return callback(hash.digest(digestEncoding));
+    }
+
+    if (body.mode === RequestBody.MODES.urlencoded) {
+        urlencodedBody = rfc3986(querystring.stringify(urlencodedBodyBuilder(body.urlencoded).form));
```

Comment:
too convoluted

Reply:
This will be removed once we use postman-url-encoder directly here. Will do that while making changes related to encoding in different PR.